### ISSUE TITLE
Adding a pixel shader to Atlas engine explodes

### ISSUE THREAD

**Issue by zadjii-msft** [microsoft/terminal#13899]
On `main` + Dev Build, so `useAtlas:true` by default.

```jsonc
            {
                "background": "#011424",
                "colorScheme": "Campbell Powershell",
                "commandline": "C:\\Program Files\\PowerShell\\7\\pwsh.exe -nologo",
                "font":
                {
                    "face": "Cascadia Code",
                    "weight": "light"
                },
                "hidden": false,
                "name": "PowerShell",
                "opacity": 50,
                "unfocusedAppearance":
                {
                    "background": "#162029",
                    "experimental.pixelShaderPath": "C:\\Users\\migrie\\dev\\public\\terminal\\samples\\PixelShaders\\Grayscale.hlsl",
                },
                "useAcrylic": true,
            },
```

(modify path as necessary)

<details>
<summary>stack</summary>

```
     Microsoft.Terminal.Control.dll!std::_Iterator_base12::_Orphan_me_unlocked_v3() Line 1204    C++
     Microsoft.Terminal.Control.dll!std::_Iterator_base12::_Orphan_me_locked_v3() Line 1217    C++
     Microsoft.Terminal.Control.dll!std::_Iterator_base12::~_Iterator_base12() Line 1125    C++
     Microsoft.Terminal.Control.dll!std::_List_unchecked_const_iterator<std::_List_val<std::_List_simple_types<std::pair<Microsoft::Console::Render::AtlasEngine::AtlasKey,Microsoft::Console::Render::AtlasEngine::AtlasValue>>>,std::_Iterator_base12>::~_List_unchecked_const_iterator<std::_List_val<std::_List_simple_types<std::pair<Microsoft::Console::Render::AtlasEngine::AtlasKey,Microsoft::Console::Render::AtlasEngine::AtlasValue>>>,std::_Iterator_base12>()    C++
     Microsoft.Terminal.Control.dll!std::_List_const_iterator<std::_List_val<std::_List_simple_types<std::pair<Microsoft::Console::Render::AtlasEngine::AtlasKey,Microsoft::Console::Render::AtlasEngine::AtlasValue>>>>::~_List_const_iterator<std::_List_val<std::_List_simple_types<std::pair<Microsoft::Console::Render::AtlasEngine::AtlasKey,Microsoft::Console::Render::AtlasEngine::AtlasValue>>>>()    C++
     Microsoft.Terminal.Control.dll!std::_List_iterator<std::_List_val<std::_List_simple_types<std::pair<Microsoft::Console::Render::AtlasEngine::AtlasKey,Microsoft::Console::Render::AtlasEngine::AtlasValue>>>>::~_List_iterator<std::_List_val<std::_List_simple_types<std::pair<Microsoft::Console::Render::AtlasEngine::AtlasKey,Microsoft::Console::Render::AtlasEngine::AtlasValue>>>>()    C++
     Microsoft.Terminal.Control.dll!std::_List_iterator<std::_List_val<std::_List_simple_types<std::pair<Microsoft::Console::Render::AtlasEngine::AtlasKey,Microsoft::Console::Render::AtlasEngine::AtlasValue>>>>::`scalar deleting destructor'(unsigned int)    C++
     Microsoft.Terminal.Control.dll!std::_Destroy_in_place<std::_List_iterator<std::_List_val<std::_List_simple_types<std::pair<Microsoft::Console::Render::AtlasEngine::AtlasKey,Microsoft::Console::Render::AtlasEngine::AtlasValue>>>>>(std::_List_iterator<std::_List_val<std::_List_simple_types<std::pair<Microsoft::Console::Render::AtlasEngine::AtlasKey,Microsoft::Console::Render::AtlasEngine::AtlasValue>>>> & _Obj) Line 302    C++
     Microsoft.Terminal.Control.dll!std::destroy_n<std::_List_iterator<std::_List_val<std::_List_simple_types<std::pair<Microsoft::Console::Render::AtlasEngine::AtlasKey,Microsoft::Console::Render::AtlasEngine::AtlasValue>>>> *,unsigned __int64>(std::_List_iterator<std::_List_val<std::_List_simple_types<std::pair<Microsoft::Console::Render::AtlasEngine::AtlasKey,Microsoft::Console::Render::AtlasEngine::AtlasValue>>>> * _First, const unsigned __int64 _Count_raw) Line 578    C++
>    Microsoft.Terminal.Control.dll!Microsoft::Console::Render::AtlasEngine::Buffer<std::_List_iterator<std::_List_val<std::_List_simple_types<std::pair<Microsoft::Console::Render::AtlasEngine::AtlasKey,Microsoft::Console::Render::AtlasEngine::AtlasValue>>>>,8>::destroy() Line 321    C++
     Microsoft.Terminal.Control.dll!Microsoft::Console::Render::AtlasEngine::Buffer<std::_List_iterator<std::_List_val<std::_List_simple_types<std::pair<Microsoft::Console::Render::AtlasEngine::AtlasKey,Microsoft::Console::Render::AtlasEngine::AtlasValue>>>>,8>::operator=(Microsoft::Console::Render::AtlasEngine::Buffer<std::_List_iterator<std::_List_val<std::_List_simple_types<std::pair<Microsoft::Console::Render::AtlasEngine::AtlasKey,Microsoft::Console::Render::AtlasEngine::AtlasValue>>>>,8> && other) Line 230    C++
     [External Code]    
     Microsoft.Terminal.Control.dll!Microsoft::Console::Render::AtlasEngine::_createResources() Line 543    C++
     Microsoft.Terminal.Control.dll!Microsoft::Console::Render::AtlasEngine::GetSwapChainHandle() Line 309    C++
     Microsoft.Terminal.Control.dll!winrt::Microsoft::Terminal::Control::implementation::ControlCore::SwapChainHandle() Line 1570    C++
     Microsoft.Terminal.Control.dll!winrt::impl::produce<winrt::Microsoft::Terminal::Control::implementation::ControlCore,winrt::Microsoft::Terminal::Control::IControlCore>::get_SwapChainHandle(unsigned __int64 * value) Line 2031    C++
     Microsoft.Terminal.Control.dll!winrt::impl::consume_Microsoft_Terminal_Control_IControlCore<winrt::Microsoft::Terminal::Control::IControlCore>::SwapChainHandle() Line 122    C++
     Microsoft.Terminal.Control.dll!winrt::Microsoft::Terminal::Control::implementation::TermControl::RenderEngineSwapChainChanged(winrt::Windows::Foundation::IInspectable __coro_heap_ellision, winrt::Windows::Foundation::IInspectable __coro_frame_ptr) Line 745    C++
```
</details>

I saw another that looked like:

<details>
<summary>Details</summary>

```
 	ucrtbased.dll!00007ffc27cf024e()	Unknown
 	Microsoft.Terminal.Control.dll!operator delete(void * block) Line 38	C++
>	Microsoft.Terminal.Control.dll!Microsoft::Console::Render::AtlasEngine::Buffer<std::_List_iterator<std::_List_val<std::_List_simple_types<std::pair<Microsoft::Console::Render::AtlasEngine::AtlasKey,Microsoft::Console::Render::AtlasEngine::AtlasValue>>>>,8>::deallocate(std::_List_iterator<std::_List_val<std::_List_simple_types<std::pair<Microsoft::Console::Render::AtlasEngine::AtlasKey,Microsoft::Console::Render::AtlasEngine::AtlasValue>>>> * data) Line 315	C++
 	Microsoft.Terminal.Control.dll!Microsoft::Console::Render::AtlasEngine::Buffer<std::_List_iterator<std::_List_val<std::_List_simple_types<std::pair<Microsoft::Console::Render::AtlasEngine::AtlasKey,Microsoft::Console::Render::AtlasEngine::AtlasValue>>>>,8>::destroy() Line 321	C++
 	Microsoft.Terminal.Control.dll!Microsoft::Console::Render::AtlasEngine::Buffer<std::_List_iterator<std::_List_val<std::_List_simple_types<std::pair<Microsoft::Console::Render::AtlasEngine::AtlasKey,Microsoft::Console::Render::AtlasEngine::AtlasValue>>>>,8>::operator=(Microsoft::Console::Render::AtlasEngine::Buffer<std::_List_iterator<std::_List_val<std::_List_simple_types<std::pair<Microsoft::Console::Render::AtlasEngine::AtlasKey,Microsoft::Console::Render::AtlasEngine::AtlasValue>>>>,8> && other) Line 230	C++
 	[External Code]	
 	Microsoft.Terminal.Control.dll!Microsoft::Console::Render::AtlasEngine::_createResources() Line 543	C++
 	Microsoft.Terminal.Control.dll!Microsoft::Console::Render::AtlasEngine::GetSwapChainHandle() Line 309	C++
 	Microsoft.Terminal.Control.dll!winrt::Microsoft::Terminal::Control::implementation::ControlCore::SwapChainHandle() Line 1570	C++
 	Microsoft.Terminal.Control.dll!winrt::impl::produce<winrt::Microsoft::Terminal::Control::implementation::ControlCore,winrt::Microsoft::Terminal::Control::IControlCore>::get_SwapChainHandle(unsigned __int64 * value) Line 2031	C++
 	Microsoft.Terminal.Control.dll!winrt::impl::consume_Microsoft_Terminal_Control_IControlCore<winrt::Microsoft::Terminal::Control::IControlCore>::SwapChainHandle() Line 122	C++
 	Microsoft.Terminal.Control.dll!winrt::Microsoft::Terminal::Control::implementation::TermControl::RenderEngineSwapChainChanged(winrt::Windows::Foundation::IInspectable __coro_heap_ellision, winrt::Windows::Foundation::IInspectable __coro_frame_ptr) Line 745	C++
 	[External Code]	
 	Microsoft.Terminal.Control.dll!wil::details::dispatcher_handler::Complete() Line 88	C++

```

</details>

looks like we're calling `Buffer::operator=()`, calling `destroy()` on `this`, and `this` was null, cause all the members of `_r` were null...

**Comment by lhecker** [comment]
I investigated this yesterday but forgot to write up my findings...
It's not just `Buffer` that's crashing - if you run it often enough, all sorts of `_r` members are causing problems. This leads me to strongly suspect a memory corruption (unlikely as `_r` isn't heap allocated), a race condition or some other kind of thread safety issue. Since `SwapChainHandle()` always appears in the stack trace and since `SwapChainHandle()` is highly thread-unsafe (doesn't take console locks, creates D3D devices if necessary concurrently with the rendering loop without memory synchronization primitives, etc.), I fixed the thread-safety issue in `SwapChainHandle()` in #13906. The issue then stopped appearing.
